grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/DeferredClientCall.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.client.grpc;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.locks.ReentrantLock;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import com.linecorp.armeria.common.CommonPools;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+
+import io.grpc.ClientCall;
+import io.grpc.Metadata;
+import io.grpc.Status;
+
+@UnstableApi
+public final class DeferredClientCall<I, O> extends ClientCall<I, O> {
+
+    private static final List<Runnable> NOOP_TASKS = ImmutableList.of();
+
+    private List<Runnable> pendingTasks = new ArrayList<>();
+
+    private final ReentrantLock lock = new ReentrantLock();
+
+    private final Executor executor;
+
+    @Nullable
+    private DeferredListner<O> deferredListner;
+
+    @Nullable
+    private volatile ArmeriaClientCall<I, O> delegate;
+
+    public DeferredClientCall(CompletableFuture<ClientCall<I, O>> clientFuture) {
+        executor = MoreExecutors.newSequentialExecutor(CommonPools.blockingTaskExecutor());
+        clientFuture.handleAsync((delegate, cause) -> {
+            if (cause != null) {
+                // TODO
```

Comment:
What do you think is the best way to notify of completion?

Reply:
We can propagate the cause via `ClientCall.cancel(...)`

Reply:
At this point it seems that the `ClientCall ` has not been generated, is there a good way to propagate the cause  via `ClientCall.cancel(...)`?

Reply:
@ikhoon 
sorry, I haven't worked on this issue recently. 
it's been a long time...
I'd like to know your opinion about this.

Reply:
Seems good!
Thanks!